### ISSUE TITLE
Question regarding reproducibility

### ISSUE THREAD

**Issue by giles200619** [stevenlsw/reart#2]
Hello authors,
Thanks very much for making the great work public! 

I noticed that the result is different after each run, I wonder if you have any tip for increasing reproducibility or if you know what is causing the result to be different for each run even when the seeds are set?  

Thank you!

**Comment by stevenlsw** [comment]
Hi, 

I think it is because of pytorch optimization. You could set `torch.backends.cudnn.benchmark = False`  and `torch.use_deterministic_algorithms(True)`, and also fix the seed. You could check the provided checkpoints and compare to your optimizations. Please let me know if you have any questions.

**Comment by giles200619** [comment]
Hello, 

Thanks very much for the quick response! 
After setting:
```
torch.backends.cudnn.deterministic = True
torch.backends.cudnn.benchmark = False
torch.use_deterministic_algorithms(True)
torch.cuda.manual_seed_all(args.manual_seed)
torch.manual_seed(args.manual_seed)
np.random.seed(args.manual_seed)
random.seed(args.manual_seed)
```
I'm still getting different eval numbers after two runs, do you have an idea of what module might be causing the randomness? 
Could it be something in the flow loss or assignment loss? 

Thank you!

**Comment by stevenlsw** [comment]
Hi, 

in flow loss and also KNN library, there could be uncertainty. Given it is optimization approaches, you might get slight different results because of platform and environment. We also use **Sec 3.33** Canonical frame selection to find the suitable model. You could first test the provided pretrained models and compared the performance. After the model is trained, the evaluation should be stable.

**Comment by giles200619** [comment]
Yes I found that the knn backward is non-deterministic, thanks very much!
